Update how Dynamo decides to graph break on an OpOverloadPacket

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112200

Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When Dynamo is about to write a call to an operator to the FX graph,
we check if it is PT2 compliant.
- For OpOverload, we check to see if the tag is on it
- For OpOverloadPacket, we do overload resolution and check to see if
  the tag is on the OpOverload that it resolves to.

Test Plan:
- new tests, existing tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

zou3519 committed:
Update how Dynamo decides to graph break on an OpOverloadPacket

Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When the operator is invoked (via call_function), we first check if
it is an OpOverload or an OpOverloadPacket. If it is an OpOverloadPacket
with a single overload, then we check the pt2_compliant_tag.
- If the OpOverloadPacket has multiple overloads, then we attempt to
grab the fake values on the node that we create and then use those to
resolve the overload. Then, we check if the resolved overload has the
pt2_compliant_tag

Test Plan:
- new tests, existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Update how Dynamo decides to graph break on an OpOverloadPacket"


Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When the operator is invoked (via call_function), we first check if
it is an OpOverload or an OpOverloadPacket. If it is an OpOverloadPacket
with a single overload, then we check the pt2_compliant_tag.
- If the OpOverloadPacket has multiple overloads, then we attempt to
grab the fake values on the node that we create and then use those to
resolve the overload. Then, we check if the resolved overload has the
pt2_compliant_tag

Test Plan:
- new tests, existing tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "Update how Dynamo decides to graph break on an OpOverloadPacket"


Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When the operator is invoked (via call_function), we first check if
it is an OpOverload or an OpOverloadPacket. If it is an OpOverloadPacket
with a single overload, then we check the pt2_compliant_tag.
- If the OpOverloadPacket has multiple overloads, then we attempt to
grab the fake values on the node that we create and then use those to
resolve the overload. Then, we check if the resolved overload has the
pt2_compliant_tag

Test Plan:
- new tests, existing tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "Update how Dynamo decides to graph break on an OpOverloadPacket"


Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When the operator is invoked (via call_function), we first check if
it is an OpOverload or an OpOverloadPacket. If it is an OpOverloadPacket
with a single overload, then we check the pt2_compliant_tag.
- If the OpOverloadPacket has multiple overloads, then we attempt to
grab the fake values on the node that we create and then use those to
resolve the overload. Then, we check if the resolved overload has the
pt2_compliant_tag

Test Plan:
- new tests, existing tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/variables/torch.py:
```diff
@@ -272,6 +293,7 @@ def call_function(
         constant_args = check_constant_args(args, kwargs)
         unspec_python_args = check_unspec_python_args(args, kwargs)
         options = VariableTracker.propagate(self, args, kwargs.values())
+        check_allowed_op(self.value)
```
One thing I'm a bit confused about:

(1) why do `check_allowed_op` and `check_allowed_overload_packet` get run in different places?

(2) It seems like `check_allowed_op` also has special handling for when it sees an OverloadPacket that only contains 1 overload? Why isn't that already handled by `check_allowed_overload_packet`?

zou3519 replied:
TL;DR: I'm doing some defensive programming (one of these callsites is much more robust than the other, so we prefer to put as much as we can there)

> (1) why do check_allowed_op and check_allowed_overload_packet get run in different places?

check_allowed_op runs at the beginning of TorchVariable.call_function.
check_allowed_overload_packet runs somewhere in the middle of TorchVariable.call_function.

They run at different places because:
- The beginning of TorchVariable.call_function is a robust place to put the check: if the TorchVariable is an opoverload, we check the tag and graph break accordingly.
- check_allowed_overload_packet requires the sample inputs to do overload resolution. So the check needs to be somewhere in the middle of TorchVariable.call_function where we can actually acquire the sample inputs.
- The middle of TorchVariable.call_function is not the most robust place because there is a large chain of if-then-else. I'm worried someone will come along and add a special case for OpOverloadPacket that bypasses the logic. There are already special cases for things like torch.ops.aten.sym_sizes in the middle that bypass this logic (but we already tag all aten ops with the pt2_compliant_tag, so these special cases are OK).

> (2) It seems like check_allowed_op also has special handling for when it sees an OverloadPacket that only contains 1 overload? Why isn't that already handled by check_allowed_overload_packet?

Because most custom ops only have 1 overload, we can handle this case without actually grabbing the sample inputs. check_allowed_op is the more robust check due to where it is located in the code, so I put the check for 1 overload there.

zou3519 replied:
Something more robust is if we create a new TorchOpsVariable to represent everything in `torch.ops.*` and then refactor all the existing code to handle that (to go into TorchOpsVariable instead of TorchVariable). This makes the if-then-else a bit less brittle

zou3519 committed:
Update on "Update how Dynamo decides to graph break on an OpOverloadPacket"


Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When the operator is invoked (via call_function), we first check if
it is an OpOverload or an OpOverloadPacket. If it is an OpOverloadPacket
with a single overload, then we check the pt2_compliant_tag.
- If the OpOverloadPacket has multiple overloads, then we attempt to
grab the fake values on the node that we create and then use those to
resolve the overload. Then, we check if the resolved overload has the
pt2_compliant_tag

Test Plan:
- new tests, existing tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "Update how Dynamo decides to graph break on an OpOverloadPacket"


Previously, under config.only_allow_pt2_compliant_ops, Dynamo graph
breaks when it see an OpOverloadPacket where any overloads are not
PT2 compliant. This is potentially brittle: if someone (unlikely) adds
a new overload for a custom operator, then this would cause a
previously non-graph-breaking call to the OpOverloadPacket to graph break.

In this PR:
- When Dynamo is about to write a call to an operator to the FX graph,
we check if it is PT2 compliant.
- For OpOverload, we check to see if the tag is on it
- For OpOverloadPacket, we do overload resolution and check to see if
  the tag is on the OpOverload that it resolves to.

Test Plan:
- new tests, existing tests

[ghstack-poisoned]

bdhirsh reviewed (approved):
Just thinking - what if we have a C++ custom autograd op, that redispatches to another custom op? In that case, dynamo will never see the second custom op and have a chance to graph break on it.

Should we add a similar version of this check in FakeTensor? (which is similar to what fake tensor does for data dependent ops)

zou3519 commented:
@pytorchbot merge -f "trying to get this internally asap"
